app/Helpers/Storage.php

```diff
@@ -52,4 +55,58 @@ public function uploadS3($presigned_url)
         ];
     }
 
+
+    /**
+     * [uploadFTP description]
+     * @param  array $FTPcredentials The credentials for the FTP server
+     * @return array                 Upload's result
+     */
+    public function uploadFTP($FTPcredentials, $invoiceId) {
+        if (ends_with($FTPcredentials["path"], "/") == false) {
+            $FTPcredentials["path"] = $FTPcredentials["path"] . "/";
+        }
+
+        $adapter = new Adapter([
+            'host' => $FTPcredentials["host"],
+            'username' => $FTPcredentials["username"],
+            'password' => $FTPcredentials["password"],
+            'port' => isset($FTPcredentials["port"]) ? $FTPcredentials["port"] : 21,
+            'root' => $FTPcredentials["path"],
+            'passive' => isset($FTPcredentials["passive"]) ? $FTPcredentials["passive"] : true,
+            'ssl' => isset($FTPcredentials["ssl"]) ? $FTPcredentials["ssl"] : false,
+            'timeout' => 300
+        ]);
+
+        try {
+            $adapter->getConnection();
+        } catch (\RuntimeException $e) {
+            return [
+                'uploaded' => false,
+                'message' => $e->getMessage(),
+                'path' => null,
+            ];
+        }
+
+        $filesystem = new Filesystem($adapter);
+
+        $filePath = "invoice_". $invoiceId ."_" . time() .".pdf";
```

Comment:
Can you move this line into `InvoiceController` please ?

Reply:
You mean this `` $filePath = "invoice_". $invoiceId ."_" . time() .".pdf";`` ?
To have the filename like a parameter for the ``uploadFTP`` method?

Reply:
yes !

if we add a new storage method later, i would be nice to reuse that line